util/cloud/aws.cc

```diff
@@ -362,6 +363,31 @@ bool IsExpiredBody(string_view body) {
   return body.find(kExpiredTokenSentinel) != std::string::npos;
 }
 
+bool AwsIsEscaped(char c) {
+  return !((c >= 'A' && c <= 'Z') ||
+    (c >= 'a' && c <= 'z') ||
+    (c >= '0' && c <= '9') ||
+    c == '-' ||
+    c == '.' ||
+    c == '_' ||
+    c == '~'
+  );
+}
+
+// Escapes the given path as documented by
+// https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html.
+std::string AwsEscapePath(const std::string_view& path, bool encode_sep) {
+  std::string escaped;
+  for (char c : path) {
+    if (!AwsIsEscaped(c) || (c == '/' && !encode_sep)) {
+      escaped.push_back(c);
+    } else {
+      escaped += absl::StrFormat("%%%02X", c);
```

Comment:
nit, this allocates a string for every escaped character.
you can achieve the same with `absl::StrAppend(&escaped, "%%",  absl::Hex(c, absl::kZeroPad2))`

Reply:
Trying that `absl::Hex` returns lowercase though we need uppercase to match the signature